deep_quoridor/src/agents/greedy.py

```diff
@@ -0,0 +1,157 @@
+        first_moves = self._valid_pawn_movements(game, observation["action_mask"])
+        shortest_path = None
+
+        # Find the shortest path for each of the first moves and keep the shortest one.
+        for move in first_moves:
+            path = self._shortest_path_from(game, move, target_row)
+            if shortest_path is None or len(path) < len(shortest_path):
+                shortest_path = [move] + path
+
+        assert shortest_path is not None, "No path found"
+        return shortest_path
+
+    def _get_opponent_position(self, game, board: np.ndarray) -> Position:
+        coords = tuple(int(x[0]) for x in np.where(board == 2))
+        assert len(coords) == 2, "Expected exactly one opponent position"
+        return coords
+
+    def _get_block_action(self, game, opponent_shortest_path: list[Position], action_mask: np.ndarray):
+        for p0, p1 in zip(opponent_shortest_path, opponent_shortest_path[1:]):
+            # For this movement of the opponent, get the possible wall placements
+            if p0[0] == p1[0]:
+                # Horizontal movement
+                rows = [p0[0], p0[0] - 1]
+                cols = [min(p0[1], p1[1])]
+                orientation = 1
+            elif p0[1] == p1[1]:
+                # Vertical movement
+                rows = [min(p0[0], p1[0])]
+                cols = [p0[1], p0[1] - 1]
+                orientation = 2
+            else:
+                assert "Expected horizontal or vertical movement"
+
+            # Check if any of the wall placements are valid.
+            # TO DO: instead of just returning the first, we can check which one is better
+            # by looking into which makes the difference of the distances more favorable
+            for r in rows:
+                for c in cols:
+                    if r < 0 or c < 0 or r >= game.wall_size or c >= game.wall_size:
+                        continue
+
+                    action = game.action_params_to_index(r, c, orientation)
+                    if action_mask[action] == 1:
+                        return action
+        # No actions found
+        return None
+
+    def get_action(self, game):
+        observation, _, termination, truncation, _ = game.last()
+        if termination or truncation:
+            return None
+
+        goal_row = game.board_size - 1 if self.player_id == "player_0" else 0
+        opponent_row = game.board_size - 1 - goal_row
+
+        my_shortest_path = self._shortest_path_from_me(game, observation, goal_row)
+        opponent_pos = self._get_opponent_position(game, observation["observation"]["board"])
+        opponent_shortest_path = self._shortest_path_from(game, opponent_pos, opponent_row)
+
+        # TODO: use a more elaborate logic to decide whether to block, could be probabilistic
+        block = False
+        if len(opponent_shortest_path) < len(my_shortest_path):
+            if len(opponent_shortest_path) < 5:
```

Comment:
One suggestion for a slightly fancier way to do this in the future would be to compute _both_ the  best block action and compute the shortest path action, then choose between those based on which one leads to the highest `(len(opponent_shortest_path) - len(my_shortest_path)`.

Reply:
yeah, good idea, sometimes there may be a great opportunity for blocking that we're missing because the condition is not met